src/main/java/com/google/devtools/build/lib/runtime/BlazeRuntime.java

```diff
@@ -160,6 +160,7 @@
   private final String productName;
   private final BuildEventArtifactUploaderFactoryMap buildEventArtifactUploaderFactoryMap;
   private final ActionKeyContext actionKeyContext;
+  private final ImmutableMap<String, AuthHeadersProvider> authHeadersProviderMap;
```

Comment:
are these values mutable?

Reply:
No for some definition of mutable. The internal state of a AuthHeadersProvider may change over time after i.e. a call to `refresh()`

Reply:
I'm confused by this. It seems kinda mutable if you need to call refresh() on it periodically. Is it thread-safe? Does it matter if different clients racily call refresh() on it?

Reply:
Is it the `ImmutableMap` that confuses you? I am happy to change it to a normal Map.

Reply:
ImmutableMap is great. The values in the maps being (pseudo?) mutable is more of a concern. How do multiple calls to the AuthHeadersProvider interact?

Reply:
Maybe this is a bit too abstract. Let me give you an example: How the Google Cloud Authentication will play out in practice is that clients (BES, remote execution) will call `getRequestHeaders` to get http headers that they will want to add to their requests in order for them to be authenticated. Once an hour or so `getRequestHeaders` will notice that its token is about to expire and transparently do a call to request a new one. So `getRequestHeaders` will return the same value for an hour.

>  How do multiple calls to the AuthHeadersProvider interact?

`AuthHeadersProvider` implementations are thread-safe and the same object can be used across independent corners of the code. One caller won't be able to observer a different behavior whether he is using the object exclusively or shared with other callers. So while it's not immutable, it's safe to share it across Bazel.

Reply:
thanks!